rts/javascript/values.js

```diff
@@ -0,0 +1,205 @@
+  if (num_dim == 0) {
+    return read_bin_scalar(buff, typ);
+  } else {
+    return read_bin_array(buff, num_dim, typ);
+  }
+}
+
+var typToSize = {
+  "bool" : 1,
+  "  u8" : 1,
+  "  i8" : 1,
+  " u16" : 2,
+  " i16" : 2,
+  " u32" : 4,
+  " i32" : 4,
+  " f32" : 4,
+  " u64" : 8,
+  " i64" : 8,
+  " f64" : 8,
+}
+
+function toU8(ta) {
+  return new Uint8Array(ta.buffer, ta.byteOffset, ta.byteLength);
+}
+
+
+function construct_binary_value(v, typ) {
+  var dims;
+  var payload_bytes;
+  var filler;
+  if (v instanceof FutharkOpaque) {
+    throw "Opaques are not supported";
+  } else if (v instanceof FutharkArray) {
+    var t = v.futharkType();
+    var ftype = "    ".slice(t.length) + t;
+    var shape = v.shape();
+    var ta = v.toTypedArray(shape);
+    var da = new BigInt64Array(shape);
+    dims = shape.length;
+    payload_bytes = da.byteLength + ta.byteLength;
+    filler = (bytes) => {
+      bytes.set(toU8(da), 7);
+      bytes.set(toU8(ta), 7 + da.byteLength);
+    }
+  } else {
+    var ftype = "    ".slice(typ.length) + typ;
+    dims = 0;
+    payload_bytes = typToSize[ftype];
+    filler = (bytes) => {
+      var scalar = new (typToType[ftype])([v]);
+      bytes.set(toU8(scalar), 7);
+    }
+  }
+  var total_bytes = 7 + payload_bytes;
+  var bytes = new Uint8Array(total_bytes);
+  bytes[0] = Buffer.from('b').readUInt8();
+  bytes[1] = 2; // Not sure why this
+  bytes[2] = dims;
+  for (var i = 0; i < 4; i++) {
+    bytes[3+i] = ftype.charCodeAt(i);
+  }
+  filler(bytes);
+  return Buffer.from(bytes);
+}
+
+class Reader {
+  constructor(f) {
+    this.f = f;
+    // TODO update this buff when value is read
+    this.buff = fileToBuff(f);
+    this.old_buff = this.buff;
+  }
+
+  read_bin_array(num_dim, typ) {
+    var u8_array = new Uint8Array(num_dim * 8);
+    binToStringArray(this.buff.slice(0, num_dim * 8), u8_array);
+    var shape = new BigInt64Array(u8_array.buffer);
+    var length = shape[0];
+    for (var i = 1; i < shape.length; i++) {
+      length = length * shape[i];
+    }
+    length = Number(length);
+    var dbytes = typToSize[typ];
+    var u8_data = new Uint8Array(length * dbytes);
+    binToStringArray(this.buff.slice(num_dim * 8, num_dim * 8 + dbytes * length), u8_data);
+    var data  = new (typToType[typ])(u8_data.buffer);
+    var tmp_buff = this.buff.slice(num_dim * 8, num_dim * 8 + dbytes * length);
+    this.buff = this.buff.slice(num_dim * 8 + dbytes * length);
+   return [shape, data];
+  }
+
+  read_bin_scalar(typ) {
+    var size = typToSize[typ];
+    var u8_array = new Uint8Array(size);
+    binToStringArray(this.buff, u8_array);
+    var array = new (typToType[typ])(u8_array.buffer);
+    this.buff = this.buff.slice(u8_array.length); // Update buff to be unread part of the string
+    return array[0];
+  }
+
+  skip_spaces() {
+    while (this.buff.length > 0 && this.buff.slice(0, 1).toString().trim() == "") {
+      this.buff = this.buff.slice(1);
+    }
+  }
+
+  read_binary() {
```

Comment:
How does this differ from the `read_binary` on line 46?

Reply:
The other one was actually some old code that was no longer used. Good catch